### PR TITLE
LDAP improvement & bug fix

### PR DESCRIPTION
**Pull request by rkorn86**
The following PR addresses some LDAP bugs and improvements.

1. Compare custom user password LDAP-Attribute  
  1.1 New property `cas.authn.ldap[0].principalAttributePassword` which is by default  set to `userPassword`  
   1.2 side effect: `AUTHENTICATED` and `ANONYMOUS` ldap authentication dont use `PooledBindAuthenticationHandler` anymore, but compare LDAP-PasswordAttribute (which needs be readable via ACL of course)
2. Fix LDAP connection Pool not initialized error
3. Initialize LdapAuthenticationHandler correctly

Ensure that you include the following:

- [x] Brief description of changes applied
- [x] Any documentation on how to configure, test
- [x] Any possible limitations, side effects, etc
- [x] Reference any other pull requests that might be related.
- [x] I have signed the Apereo ICLA form, located here: https://www.apereo.org/licensing/agreements/icla

### CODE REVIEW

**Comment by rkorn86** [comment]
I wont touch switch statement in `Beans.java`, as this style bug is out of this PR scope.

File: cas-server-support-ldap/src/main/java/org/apereo/cas/config/LdapAuthenticationConfiguration.java [hunk 1]
```diff
@@ -173,5 +171,11 @@ private static PooledBindAuthenticationHandler getPooledBindAuthenticationHandle
         handler.setAuthenticationControls(new PasswordPolicyControl());
         return handler;
     }
-}
 
+    private static PooledCompareAuthenticationHandler getPooledCustomCompareAuthenticationHandler(final LdapAuthenticationProperties l) {
+        final PooledCompareCustomAttributeAuthenticationHandler handler = new PooledCompareCustomAttributeAuthenticationHandler(
+                Beans.newPooledConnectionFactory(l));
+        handler.setPasswordAttributeName(l.getPrincipalAttributePassword());
+        return handler;
```
Comment (mmoayyed) [inline cas-server-support-ldap/src/main/java/org/apereo/cas/config/LdapAuthenticationConfiguration.java]: If the principal attribute password is not defined, could we fallback to the pooledBindAuthnHandler?

File: cas-server-support-ldap/src/main/java/org/apereo/cas/authentication/PooledCompareCustomAttributeAuthenticationHandler.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import org.ldaptive.pool.PooledConnectionFactory;
+
+/**
+ * Extended ldaptive {@code Authenticator} component, which uses a custom ldap
+ * attribute for password comparison.
+ *
+ * @author Robert Kornmesser
+ * @since 5.0.0
+ */
+public class PooledCompareCustomAttributeAuthenticationHandler extends PooledCompareAuthenticationHandler {
```
Comment (mmoayyed) [inline cas-server-support-ldap/src/main/java/org/apereo/cas/authentication/PooledCompareCustomAttributeAuthenticationHandler.java]: Could this be something we could actually contribute to ldaptive? It doesnt seem like something we'd want to maintain. Have you discussed this change with Daniel @ ldaptive?

File: cas-server-core-configuration/src/main/java/org/apereo/cas/configuration/model/support/ldap/LdapAuthenticationProperties.java [hunk 3]
```diff
@@ -40,6 +40,7 @@
     private boolean usePasswordPolicy = true;
     private String dnFormat;
     private String principalAttributeId = "uid";
+    private String principalAttributePassword = "userPassword";
```
Comment (mmoayyed) [inline cas-server-core-configuration/src/main/java/org/apereo/cas/configuration/model/support/ldap/LdapAuthenticationProperties.java]: You'll need to also document this property, and briefly explain what its purpose it perhaps in the docs where properties are all listed.

File: cas-server-support-ldap/src/main/java/org/apereo/cas/config/LdapAuthenticationConfiguration.java (hunk 1 above)
Comment (rkorn86) [inline cas-server-support-ldap/src/main/java/org/apereo/cas/config/LdapAuthenticationConfiguration.java]: Dont know. Should we?

File: cas-server-support-ldap/src/main/java/org/apereo/cas/authentication/PooledCompareCustomAttributeAuthenticationHandler.java (hunk 2 above)
Comment (rkorn86) [inline cas-server-support-ldap/src/main/java/org/apereo/cas/authentication/PooledCompareCustomAttributeAuthenticationHandler.java]: Did not have closer look on ldaptive. Maybe that is already integrated in a higher version, which we dont use yet. I see a PR for ldaptive 1.2. Maybe we should update ldaptive before?

**Comment by mmoayyed** [comment]
@rkorn86 in light of M3 coming up, do you think you'd get a chance to revisit this PR some time today,  by Monday EOD?

**Comment by rkorn86** [comment]
Dont think so, im on weekend already. My TZ is Europe/Berlin ;-)
Im back in office on monday.


Am 15. Juli 2016 5:49:31 nachm. schrieb Misagh Moayyed 
<notifications@github.com>:

**Comment by mmoayyed** [comment]
Gotcha. What might be reasonable timeline for you to review this next weekend? Alternatively, we could proceed with M3 and review this afterwards but I don't see a whole lot of sense in that. 

schönes Wochenende!

**Comment by rkorn86** [comment]
I will have a closer look at it next week, but I dont know how fast the 
ldaptive guys are.

Danke schön! :-)


Am 15. Juli 2016 5:58:36 nachm. schrieb Misagh Moayyed 
<notifications@github.com>:

File: cas-server-support-ldap/src/main/java/org/apereo/cas/config/LdapAuthenticationConfiguration.java (hunk 1 above)
Comment (mmoayyed) [inline cas-server-support-ldap/src/main/java/org/apereo/cas/config/LdapAuthenticationConfiguration.java]: I think we should, yes.

File: cas-server-support-ldap/src/main/java/org/apereo/cas/authentication/PooledCompareCustomAttributeAuthenticationHandler.java (hunk 2 above)
Comment (mmoayyed) [inline cas-server-support-ldap/src/main/java/org/apereo/cas/authentication/PooledCompareCustomAttributeAuthenticationHandler.java]: I looked at 1.2 and I don't see it. I suggest you discuss this with Daniel and see if he'd be willing to take the contribution. If that works, we can simply just upgrade to 1.2 via that other PR, and accept your change afterwards. If not, we'll merge this PR as is.

Comment (rkorn86) [inline cas-server-support-ldap/src/main/java/org/apereo/cas/authentication/PooledCompareCustomAttributeAuthenticationHandler.java]: Opened issue https://github.com/vt-middleware/ldaptive/issues/111 and created PR. How long should we wait for answer?

Comment (mmoayyed) [inline cas-server-support-ldap/src/main/java/org/apereo/cas/authentication/PooledCompareCustomAttributeAuthenticationHandler.java]: By this weekend.

Comment (mmoayyed) [inline cas-server-support-ldap/src/main/java/org/apereo/cas/authentication/PooledCompareCustomAttributeAuthenticationHandler.java]: You also want to make sure other comments on this pull are accounted for.

**Comment by rkorn86** [comment]
PR https://github.com/vt-middleware/ldaptive/pull/112 was successfully merged, so i merged the https://github.com/Unicon/cas/tree/ldaptive1.2 branch into mine. Result is, that this PR is the same as https://github.com/apereo/cas/pull/1510 + usage of the custom ldap password attribute.

**Comment by mmoayyed** [comment]
I am not sure when ldaptive 1.2 comes out. Given that, I'll give this a few more days to see if we can find a better answer. The danger here is, if we merge and ldaptive doesnt release for another 3 months then that delays the 5 release because we'd rather not ship with a snapshot. So lets wait a bit more.

Very much appreciate the work you have done here.

**Comment by mmoayyed** [comment]
@rkorn86 could you merge with master once please?

**Comment by rkorn86** [comment]
I will do today evening.


Am 22. Juli 2016 11:07:05 vorm. schrieb Misagh Moayyed 
<notifications@github.com>:

**Comment by mmoayyed** [comment]
Appreciate it. This should make it JIT for M3. Vielen Dank.

**Comment by rkorn86** [comment]
Last merge done. No Problem!:+1:

**Comment by mmoayyed** [comment]
Perfect. I'll let travis do its thing and then proceed. Thanks again.

**Pull request merged**